contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,870 @@
+        sr.grantRole(sr.REPORT_REWARDS_MINTED_ROLE(), address(_lido));
+    }
+
+    function _initializeValidatorsExitBus() internal {
+        IValidatorsExitBusOracle vebo = _validatorsExitBusOracle;
+        // NB: Setting same initial epoch as for AccountingOracle on purpose
+        _hashConsensusForValidatorsExitBusOracle.updateInitialEpoch(
+            _calcInitialEpochForAccountingOracleHashConsensus()
+        );
+        vebo.initialize(
+            address(this),
+            address(_hashConsensusForValidatorsExitBusOracle),
+            _validatorsExitBusOracleConsensusVersion,
+            VEBO_LAST_PROCESSING_REF_SLOT
+        );
+        vebo.grantRole(vebo.PAUSE_ROLE(), _gateSeal);
+        _resumeValidatorsExitBusOracle();
+    }
+
+    function _migrateLidoOracleCommitteeMembers() internal {
+        address[] memory members = _lidoOracle.getOracleMembers();
+        uint256 quorum = _lidoOracle.getQuorum();
+        bytes32 manage_members_role = _hashConsensusForAccountingOracle.MANAGE_MEMBERS_AND_QUORUM_ROLE();
+
+        _hashConsensusForAccountingOracle.grantRole(manage_members_role, address(this));
+        _hashConsensusForValidatorsExitBusOracle.grantRole(manage_members_role, address(this));
+        for (uint256 i; i < members.length; ++i) {
+            _hashConsensusForAccountingOracle.addMember(members[i], quorum);
+            _hashConsensusForValidatorsExitBusOracle.addMember(members[i], quorum);
+        }
+        _hashConsensusForAccountingOracle.renounceRole(manage_members_role, address(this));
+        _hashConsensusForValidatorsExitBusOracle.renounceRole(manage_members_role, address(this));
+    }
+
+    function _migrateDSMGuardians() internal {
+        IDepositSecurityModule previousDSM = IDepositSecurityModule(_previousDepositSecurityModule);
+        address[] memory guardians = previousDSM.getGuardians();
+        uint256 quorum = previousDSM.getGuardianQuorum();
+        _depositSecurityModule.addGuardians(guardians, quorum);
+    }
+
+    function _finishUpgrade() internal {
+        if (msg.sender != _voting) revert OnlyVotingCanUpgrade();
+        if (!isUpgradeStarted) revert StartMustBeCalledBeforeFinish();
+        if (isUpgradeFinished) revert CanOnlyFinishOnce();
+        /// Here we check that the contract got new ABI function getContractVersion(), although it is 0 yet
+        /// because in the new contract version is stored in a different slot
+        if (_legacyOracle.getContractVersion() != NOT_INITIALIZED_CONTRACT_VERSION) {
+            revert LidoOracleMustBeUpgradedToLegacy();
+        }
+        isUpgradeFinished = true;
+
+        _legacyOracle.finalizeUpgrade_v4(address(_accountingOracle));
+
+        _lido.finalizeUpgrade_v2(address(_locator), _eip712StETH);
+
+        _nodeOperatorsRegistry.finalizeUpgrade_v2(
+            address(_locator),
+            _nodeOperatorsRegistryStakingModuleType,
+            _nodeOperatorsRegistryStuckPenaltyDelay
+        );
```

Comment:
worth having asserts for the proper impl addresses also (said above)

Reply:
done